### DISCUSSION TITLE
Small File Optimize: Why is the maximum setting for MinShardSize 2KB, not is 128KB?

### DISCUSSION THREAD

**Discussion by jianwei1216** [cubefs/cubefs#3429] [Q&A] [answered]
Background 
- https://cubefs.io/docs/master/design/blobstore.html#small-file-optimization
- For example, assuming the Blob threshold is 4M, the erasure code mode is "4+4", and the Shard threshold is 1M. There is currently 128K of user data. If split in the conventional way, 128K is split into 4 parts, each part is 32K:
- ![图片](https://github.com/cubefs/cubefs/assets/19184136/1f9cfde3-040f-4b69-92b0-b1c957d4ac34)
- This slicing method requires at least 4 network requests for multiple IOs when the user reads 128K data, and the probability of exceptions is higher.
- Our solution uses space to trade time, requiring only one network request to obtain complete data. All 128K user data is written to the first data block, and the remaining three data blocks are replaced with blank data. When downloading, as long as the useful data block or any check block can be correctly obtained, all user data can be repaired.
- ![图片](https://github.com/cubefs/cubefs/assets/19184136/bd549167-c4c9-47bb-b585-818142a13480)

After analyzing the source code
- I found that the maximum value of MinShardSize is 2KB

```
type Tactic struct {
	N int
	M int
	// local parity count
	L int
	// the count of AZ, access use this for split data shards and parity shards
	AZCount int

	// PutQuorum write quorum,
	// MUST make sure that ec data is recoverable if one AZ was down
	// We SHOULD ignore the local shards
	// (N + M) / AZCount + N <= PutQuorum <= M + N
	PutQuorum int

	// get quorum config
	GetQuorum int

	// MinShardSize min size per shard, fill data into shards 0-N continuously,
	// align with zero bytes if data size less than MinShardSize*N
	//
	// length of data less than MinShardSize*N, size of per shard = MinShardSize
	//  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
	//  |  data  |                 align zero bytes                     |
	//  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
	//  |    0    |    1    |    2    |   ....                |    N    |
	//  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
	//
	// length of data more than MinShardSize*N, size of per shard = ceil(len(data)/N)
	//  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
	//  |                           data                        |padding|
	//  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
	//  |    0    |    1    |    2    |   ....                |    N    |
	//  - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
	MinShardSize int
}
```

```
// Note: Don't modify it unless you know very well how codemode works.
const (
	// align size per shard
	alignSize0B   = 0    // 0B
	alignSize512B = 512  // 512B
	alignSize2KB  = 2048 // 2KB
)

// The tactic is fixed pairing with one codemode.
// Add a new codemode if you want other features.
var constCodeModeTactic = map[CodeMode]Tactic{
	// three az
	EC15P12: {N: 15, M: 12, L: 0, AZCount: 3, PutQuorum: 24, GetQuorum: 0, MinShardSize: alignSize2KB}, //2*15=30KB
	EC6P6:   {N: 6, M: 6, L: 0, AZCount: 3, PutQuorum: 11, GetQuorum: 0, MinShardSize: alignSize2KB},  //2*6=12KB
	EC12P9:  {N: 12, M: 9, L: 0, AZCount: 3, PutQuorum: 20, GetQuorum: 0, MinShardSize: alignSize2KB}, //2*12=24KB

	// two az
	EC16P20L2: {N: 16, M: 20, L: 2, AZCount: 2, PutQuorum: 34, GetQuorum: 0, MinShardSize: alignSize2KB}, //2*16=32KB
	EC6P10L2:  {N: 6, M: 10, L: 2, AZCount: 2, PutQuorum: 14, GetQuorum: 0, MinShardSize: alignSize2KB}, //2*6=12KB

	// single az
	EC12P4: {N: 12, M: 4, L: 0, AZCount: 1, PutQuorum: 15, GetQuorum: 0, MinShardSize: alignSize2KB}, //2*12=24KB
	EC16P4: {N: 16, M: 4, L: 0, AZCount: 1, PutQuorum: 19, GetQuorum: 0, MinShardSize: alignSize2KB}, //2*16=32KB
	EC3P3:  {N: 3, M: 3, L: 0, AZCount: 1, PutQuorum: 5, GetQuorum: 0, MinShardSize: alignSize2KB},    //2*3 = 6KB
	EC10P4: {N: 10, M: 4, L: 0, AZCount: 1, PutQuorum: 13, GetQuorum: 0, MinShardSize: alignSize2KB}, //2*10=20KB
	EC6P3:  {N: 6, M: 3, L: 0, AZCount: 1, PutQuorum: 8, GetQuorum: 0, MinShardSize: alignSize2KB},    //2*6=12KB
	// for env test
	EC6P3L3:       {N: 6, M: 3, L: 3, AZCount: 3, PutQuorum: 9, GetQuorum: 0, MinShardSize: alignSize2KB}, //2*6=12KB
	EC6P6Align0:   {N: 6, M: 6, L: 0, AZCount: 3, PutQuorum: 11, GetQuorum: 0, MinShardSize: alignSize0B},
	EC6P6Align512: {N: 6, M: 6, L: 0, AZCount: 3, PutQuorum: 11, GetQuorum: 0, MinShardSize: alignSize512B}, 
	EC4P4L2:       {N: 4, M: 4, L: 2, AZCount: 2, PutQuorum: 6, GetQuorum: 0, MinShardSize: alignSize2KB}, //2*4=8KB
	EC6P6L9:       {N: 6, M: 6, L: 9, AZCount: 3, PutQuorum: 11, GetQuorum: 0, MinShardSize: alignSize2KB}, //2*6=12KB
	EC6P8L10:      {N: 6, M: 8, L: 10, AZCount: 2, PutQuorum: 13, GetQuorum: 0, MinShardSize: alignSize0B},
}
```

I am a little confused about the definition of small file size. 
For example, files such as 6K, 8K, 12K, 20K, 24K, 30K, and 32K are already very small and will consume a lot of IOPS of the HDD disk.

Q1：Why not set MinShardSize to 128KB?

Since we have chosen the optimization strategy of exchanging space for time,
Q2：Why choose MinShardSize 2KB as the default value for small file optimization?

I don't understand this very much. 
Q3：Are you afraid of too much space expansion?

Q4：I want to know the original intention of setting MinShardSize to 2KB

Q5：In actual production environments, for HDD disks, is MinShardSize = 2KB still beneficial?
- A 128KB file still requires N network and HDD disk interactions

Q6：Setting MinShardSize=2KB will take up more metadata for chunks
- For a 16GB chunk, the metadata size of MinShardSize 2KB and 128KB is 64 times different.
  - 16 * 1024 * 1024 / 2 = 8388608 
  - 16 * 1024 * 1024 / 128 = 131072

@Cloudstriff @sejust @JasonHu520 @pengtianyue025 @true1064  Can you help me? Thanks!

**Comment by sejust** [comment, marked as answer]
First of all, do not understand the definition of the value, to match the application.

The small file definition itself is strongly bound to the business. AI data scenarios are GB files, a novel site's data are in the KB level, in some cases a file is likely to be only a few bytes.

If your system is full of 100KB files, you can certainly choose to use (4,4,128KB) mode, adding 5-8 times the capacity for some read performance. (Just submit a codemode of (4, 4, 128KB) to blobstore system.)

The rationale for the base 2KB choice is to cover the scenario of storing checksum or some json data in a single file. On the 4KB block size storage side, 2KB of actual data adding the header and footer is only in one block.

**Reply by jianwei1216** [reply]
Thank you very much

**Reply by jianwei1216** [reply]
HDD 85% used：16 * 0.909 * 0.85 * 1024 = 12659GB
Chunk 16GB count: 12659 / 16 = 790 chunks
MinShardSize = 2KB

1 chunk bid count: 16 * 1024 * 1024 / 2 = 838,8608
790 chunk bid count: 8388608 * 790 = 66,2700,0320
1 bid meta size: 30 Bytes
all bid meta size: 6627000320 * 30 / 1024 / 1024 / 1024 = 185GB  (one-hdd)



According to actual calculations, 
if a system is full of relatively small files, 
such as 4KB to 128KB, I think MinShardSize = 2KB is still too dangerous, and the memory may not be enough.

The efficiency of Volume Inspect (scanning and comparison) depends on whether the bid can be fully stored in memory.

**Reply by sejust** [reply]
This the reason. If your system is full of small files, so the indexing overhead has exceeded the actual data, you should choose Nosql to store the key-value together. Or merge thousands of the small files into one file, then store it into Blobstore.

**Reply by jianwei1216** [reply]
The default value is MinShardSize=2KB.
No benefit is seen, and the objects are still small.
The minimum allocation unit of the file system is generally 4KB aligned.
This seems inconsistent with the propaganda of "trading space for time"

**Reply by jianwei1216** [reply]
> Or merge thousands of the small files into one file, then store it into Blobstore.

Garbage collection will also introduce additional read and write traffic

Is cubefs/blobstore currently performing small object aggregation based on cubefs multi-replica pool?
